src/libsyntax/parse/lexer/mod.rs

```diff
@@ -1233,26 +1233,27 @@ impl<'a> StringReader<'a> {
                        !keyword_checking_token.is_keyword(token::keywords::Static) {
                         self.err_span_(start, last_bpos, "invalid lifetime name");
                     }
+                    // lifetimes shouldn't end with a single quote
+                    // if we find one, then this is an invalid character literal
+                    if self.curr_is('\'') {
+                        panic!(self.fatal_span_verbose(
+                               // Byte offsetting here is okay because the
+                               // character before position `start` is an
+                               // ascii single quote.
+                               start - BytePos(1), self.pos,
+                               String::from("character literal may only contain one codepoint")));
```

Comment:
Probably the better approach would be to add something along the lines of `let start_with_quote = self.last_pos;` before the `self.bump` above instead of doing manual byte arithmetic here.

Also it probably would be a good idea to move check to right after the `while ident_continue` loop, so we didn’t do all the unecessary work related to lifetimes?